craft_application/application.py

```diff
@@ -0,0 +1,273 @@
+# This file is part of craft_application.
+#
+# Copyright 2023 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License version 3, as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+# SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License along
+# with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Main application classes for a craft-application."""
+from __future__ import annotations
+
+import functools
+import os
+import pathlib
+import signal
+import subprocess
+import sys
+from dataclasses import dataclass, field
+from importlib import metadata
+from typing import TYPE_CHECKING, Any, cast, final
+
+import craft_cli
+import craft_providers
+from xdg.BaseDirectory import save_cache_path  # type: ignore[import]
+
+from craft_application import commands, models
+
+if TYPE_CHECKING:
+    from craft_application.services import service_factory
+
+GLOBAL_VERSION = craft_cli.GlobalArgument(
+    "version", "flag", "-V", "--version", "Show the application version and exit"
+)
+
+
+@final
+@dataclass(frozen=True)
+class AppMetadata:
+    """Metadata about a *craft application."""
+
+    name: str
+    summary: str | None = None
```

Comment:
What's the target minimal python version? I think `A|B` is 3.10+?

Reply:
[`from __future__ import annotations`](https://docs.python.org/3.8/library/__future__.html#__future__.annotations) brings it into Python 3.7+ Combine with ruff's `UP007` rule and it'll automatically convert once you use postponed annotations (which I originally included for return values).